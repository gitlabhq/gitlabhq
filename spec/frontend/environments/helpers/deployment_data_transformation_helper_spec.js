import {
  getAuthorFromCommit,
  getCommitFromDeploymentNode,
  convertToDeploymentTableRow,
} from '~/environments/helpers/deployment_data_transformation_helper';

describe('deployment_data_transformation_helper', () => {
  const commitWithAuthor = {
    id: 'gid://gitlab/CommitPresenter/0cb48dd5deddb7632fd7c3defb16075fc6c3ca74',
    shortId: '0cb48dd5',
    message: 'Update .gitlab-ci.yml file',
    webUrl:
      'http://gdk.test:3000/gitlab-org/pipelinestest/-/commit/0cb48dd5deddb7632fd7c3defb16075fc6c3ca74',
    authorGravatar:
      'https://www.gravatar.com/avatar/e64c7d89f26bd1972efa854d13d7dd61?s=80&d=identicon',
    authorName: 'Administrator',
    authorEmail: 'admin@example.com',
    author: {
      id: 'gid://gitlab/User/1',
      name: 'Administrator',
      avatarUrl: '/uploads/-/system/user/avatar/1/avatar.png',
      webUrl: 'http://gdk.test:3000/root',
    },
  };

  const commitWithoutAuthor = {
    id: 'gid://gitlab/CommitPresenter/02274a949a88c9aef68a29685d99bd9a661a7f9b',
    shortId: '02274a94',
    message: 'Commit message',
    webUrl:
      'http://gdk.test:3000/gitlab-org/pipelinestest/-/commit/02274a949a88c9aef68a29685d99bd9a661a7f9b',
    authorGravatar:
      'https://www.gravatar.com/avatar/91811aee1dec1b2655fa56f894e9e7c9?s=80&d=identicon',
    authorName: 'Andrei Zubov',
    authorEmail: 'azubov@gitlab.com',
    author: null,
  };

  const deploymentNode = {
    id: 'gid://gitlab/Deployment/76',
    iid: '31',
    status: 'SUCCESS',
    createdAt: '2022-10-17T07:44:17Z',
    ref: 'main',
    tag: false,
    job: {
      name: 'deploy-prod',
      refName: 'main',
      id: 'gid://gitlab/Ci::Build/860',
      webPath: '/gitlab-org/pipelinestest/-/jobs/860',
      deploymentPipeline: {
        jobs: {
          nodes: [
            {
              name: 'deploy-staging',
              playable: true,
              scheduledAt: '2023-01-17T11:02:41.369Z',
              webPath: 'https://gdk.test:3000/redeploy',
            },
            {
              name: 'deploy-production',
              playable: true,
              scheduledAt: '2023-01-17T11:02:41.369Z',
              webPath: 'https://gdk.test:3000/redeploy',
            },
          ],
        },
      },
    },
    commit: commitWithAuthor,
    triggerer: {
      id: 'gid://gitlab/User/1',
      webUrl: 'http://gdk.test:3000/root',
      name: 'Administrator',
      avatarUrl: '/uploads/-/system/user/avatar/1/avatar.png',
    },
    finishedAt: '2022-10-17T07:44:43Z',
  };

  const deploymentNodeWithPipeline = {
    id: 'gid://gitlab/Deployment/76',
    iid: '31',
    status: 'SUCCESS',
    createdAt: '2022-10-17T07:44:17Z',
    ref: 'main',
    tag: false,
    job: {
      name: 'deploy-prod',
      refName: 'main',
      id: 'gid://gitlab/Ci::Bridge/860',
      webPath: null,
      deploymentPipeline: {
        id: 'gid://gitlab/Ci::Pipeline/101',
        path: '/gitlab-org/pipelinestest/-/pipelines/101',
        jobs: {
          nodes: [
            {
              name: 'deploy-staging',
              playable: true,
              scheduledAt: '2023-01-17T11:02:41.369Z',
              webPath: 'https://gdk.test:3000/redeploy',
            },
          ],
        },
      },
    },
    commit: commitWithAuthor,
    triggerer: {
      id: 'gid://gitlab/User/1',
      webUrl: 'http://gdk.test:3000/root',
      name: 'Administrator',
      avatarUrl: '/uploads/-/system/user/avatar/1/avatar.png',
    },
    finishedAt: '2022-10-17T07:44:43Z',
  };

  const deploymentNodeWithDownstreamPipeline = {
    id: 'gid://gitlab/Deployment/76',
    iid: '31',
    status: 'SUCCESS',
    createdAt: '2022-10-17T07:44:17Z',
    ref: 'main',
    tag: false,
    job: {
      name: 'deploy-prod',
      refName: 'main',
      id: 'gid://gitlab/Ci::Bridge/860',
      webPath: null,
      deploymentPipeline: {
        id: 'gid://gitlab/Ci::Pipeline/101',
        path: '/gitlab-org/pipelinestest/-/pipelines/101',
        jobs: {
          nodes: [
            {
              name: 'deploy-staging',
              playable: true,
              scheduledAt: '2023-01-17T11:02:41.369Z',
              webPath: 'https://gdk.test:3000/redeploy',
            },
          ],
        },
      },
      downStreamPipeline: {
        id: 'gid://gitlab/Ci::Pipeline/102',
        path: '/gitlab-org/pipelinestest/-/pipelines/102',
        jobs: {
          nodes: [
            {
              name: 'deploy-staging',
              playable: true,
              scheduledAt: '2023-01-17T11:02:41.369Z',
              webPath: 'https://gdk.test:3000/redeploy',
            },
          ],
        },
      },
    },
    commit: commitWithAuthor,
    triggerer: {
      id: 'gid://gitlab/User/1',
      webUrl: 'http://gdk.test:3000/root',
      name: 'Administrator',
      avatarUrl: '/uploads/-/system/user/avatar/1/avatar.png',
    },
    finishedAt: '2022-10-17T07:44:43Z',
  };

  const deploymentNodeWithNoJob = {
    ...deploymentNode,
    job: null,
    finishedAt: null,
  };

  const environment = {
    lastDeployment: {
      job: {
        name: 'deploy-production',
      },
    },
  };
  describe('getAuthorFromCommit', () => {
    it.each([commitWithAuthor, commitWithoutAuthor])('should be properly converted', (commit) => {
      expect(getAuthorFromCommit(commit)).toMatchSnapshot();
    });
  });

  describe('getCommitFromDeploymentNode', () => {
    it('should throw an error when commit field is missing', () => {
      const emptyDeploymentNode = {};

      expect(() => getCommitFromDeploymentNode(emptyDeploymentNode)).toThrow();
    });

    it('should get correclty formatted commit object', () => {
      expect(getCommitFromDeploymentNode(deploymentNode)).toMatchSnapshot();
    });
  });

  describe('convertToDeploymentTableRow', () => {
    const deploymentNodeWithEmptyJob = { ...deploymentNode, job: undefined };

    it.each([
      deploymentNode,
      deploymentNodeWithPipeline,
      deploymentNodeWithDownstreamPipeline,
      deploymentNodeWithEmptyJob,
      deploymentNodeWithNoJob,
    ])('should be converted to proper table row data', (node) => {
      expect(convertToDeploymentTableRow(node, environment)).toMatchSnapshot();
    });
  });
});
