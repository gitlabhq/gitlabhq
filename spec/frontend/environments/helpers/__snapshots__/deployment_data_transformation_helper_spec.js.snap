// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`deployment_data_transformation_helper convertToDeploymentTableRow should be converted to proper table row data 1`] = `
{
  "actions": [
    {
      "name": "deploy-staging",
      "playPath": "https://gdk.test:3000/redeploy/play",
      "playable": true,
      "scheduledAt": "2023-01-17T11:02:41.369Z",
    },
  ],
  "commit": {
    "author": {
      "avatar_url": "/uploads/-/system/user/avatar/1/avatar.png",
      "path": "http://gdk.test:3000/root",
      "username": "Administrator",
    },
    "commitRef": {
      "name": "main",
    },
    "commitUrl": "http://gdk.test:3000/gitlab-org/pipelinestest/-/commit/0cb48dd5deddb7632fd7c3defb16075fc6c3ca74",
    "shortSha": "0cb48dd5",
    "tag": false,
    "title": "Update .gitlab-ci.yml file",
  },
  "created": "2022-10-17T07:44:17Z",
  "deploymentApproval": {
    "isApprovalActionAvailable": false,
  },
  "finished": "2022-10-17T07:44:43Z",
  "id": "31",
  "job": {
    "label": "deploy-prod (#860)",
    "pipeline": undefined,
    "webPath": "/gitlab-org/pipelinestest/-/jobs/860",
  },
  "rollback": {
    "id": "gid://gitlab/Deployment/76",
    "lastDeployment": {
      "commit": {
        "author": {
          "avatarUrl": "/uploads/-/system/user/avatar/1/avatar.png",
          "id": "gid://gitlab/User/1",
          "name": "Administrator",
          "webUrl": "http://gdk.test:3000/root",
        },
        "authorEmail": "admin@example.com",
        "authorGravatar": "https://www.gravatar.com/avatar/e64c7d89f26bd1972efa854d13d7dd61?s=80&d=identicon",
        "authorName": "Administrator",
        "id": "gid://gitlab/CommitPresenter/0cb48dd5deddb7632fd7c3defb16075fc6c3ca74",
        "message": "Update .gitlab-ci.yml file",
        "shortId": "0cb48dd5",
        "webUrl": "http://gdk.test:3000/gitlab-org/pipelinestest/-/commit/0cb48dd5deddb7632fd7c3defb16075fc6c3ca74",
      },
      "isLast": false,
    },
    "name": undefined,
    "retryUrl": "/gitlab-org/pipelinestest/-/jobs/860/retry",
  },
  "status": "success",
  "triggerer": {
    "avatarUrl": "/uploads/-/system/user/avatar/1/avatar.png",
    "id": "gid://gitlab/User/1",
    "name": "Administrator",
    "webUrl": "http://gdk.test:3000/root",
  },
  "webPath": "",
}
`;

exports[`deployment_data_transformation_helper convertToDeploymentTableRow should be converted to proper table row data 2`] = `
{
  "actions": [
    {
      "name": "deploy-staging",
      "playPath": "https://gdk.test:3000/redeploy/play",
      "playable": true,
      "scheduledAt": "2023-01-17T11:02:41.369Z",
    },
  ],
  "commit": {
    "author": {
      "avatar_url": "/uploads/-/system/user/avatar/1/avatar.png",
      "path": "http://gdk.test:3000/root",
      "username": "Administrator",
    },
    "commitRef": {
      "name": "main",
    },
    "commitUrl": "http://gdk.test:3000/gitlab-org/pipelinestest/-/commit/0cb48dd5deddb7632fd7c3defb16075fc6c3ca74",
    "shortSha": "0cb48dd5",
    "tag": false,
    "title": "Update .gitlab-ci.yml file",
  },
  "created": "2022-10-17T07:44:17Z",
  "deploymentApproval": {
    "isApprovalActionAvailable": false,
  },
  "finished": "2022-10-17T07:44:43Z",
  "id": "31",
  "job": {
    "label": "deploy-prod (#860)",
    "pipeline": {
      "label": "#101",
      "path": "/gitlab-org/pipelinestest/-/pipelines/101",
    },
    "webPath": null,
  },
  "rollback": {
    "id": "gid://gitlab/Deployment/76",
    "lastDeployment": {
      "commit": {
        "author": {
          "avatarUrl": "/uploads/-/system/user/avatar/1/avatar.png",
          "id": "gid://gitlab/User/1",
          "name": "Administrator",
          "webUrl": "http://gdk.test:3000/root",
        },
        "authorEmail": "admin@example.com",
        "authorGravatar": "https://www.gravatar.com/avatar/e64c7d89f26bd1972efa854d13d7dd61?s=80&d=identicon",
        "authorName": "Administrator",
        "id": "gid://gitlab/CommitPresenter/0cb48dd5deddb7632fd7c3defb16075fc6c3ca74",
        "message": "Update .gitlab-ci.yml file",
        "shortId": "0cb48dd5",
        "webUrl": "http://gdk.test:3000/gitlab-org/pipelinestest/-/commit/0cb48dd5deddb7632fd7c3defb16075fc6c3ca74",
      },
      "isLast": false,
    },
    "name": undefined,
    "retryUrl": "null/retry",
  },
  "status": "success",
  "triggerer": {
    "avatarUrl": "/uploads/-/system/user/avatar/1/avatar.png",
    "id": "gid://gitlab/User/1",
    "name": "Administrator",
    "webUrl": "http://gdk.test:3000/root",
  },
  "webPath": "",
}
`;

exports[`deployment_data_transformation_helper convertToDeploymentTableRow should be converted to proper table row data 3`] = `
{
  "actions": [
    {
      "name": "deploy-staging",
      "playPath": "https://gdk.test:3000/redeploy/play",
      "playable": true,
      "scheduledAt": "2023-01-17T11:02:41.369Z",
    },
  ],
  "commit": {
    "author": {
      "avatar_url": "/uploads/-/system/user/avatar/1/avatar.png",
      "path": "http://gdk.test:3000/root",
      "username": "Administrator",
    },
    "commitRef": {
      "name": "main",
    },
    "commitUrl": "http://gdk.test:3000/gitlab-org/pipelinestest/-/commit/0cb48dd5deddb7632fd7c3defb16075fc6c3ca74",
    "shortSha": "0cb48dd5",
    "tag": false,
    "title": "Update .gitlab-ci.yml file",
  },
  "created": "2022-10-17T07:44:17Z",
  "deploymentApproval": {
    "isApprovalActionAvailable": false,
  },
  "finished": "2022-10-17T07:44:43Z",
  "id": "31",
  "job": {
    "label": "deploy-prod (#860)",
    "pipeline": {
      "label": "#101",
      "path": "/gitlab-org/pipelinestest/-/pipelines/101",
    },
    "webPath": null,
  },
  "rollback": {
    "id": "gid://gitlab/Deployment/76",
    "lastDeployment": {
      "commit": {
        "author": {
          "avatarUrl": "/uploads/-/system/user/avatar/1/avatar.png",
          "id": "gid://gitlab/User/1",
          "name": "Administrator",
          "webUrl": "http://gdk.test:3000/root",
        },
        "authorEmail": "admin@example.com",
        "authorGravatar": "https://www.gravatar.com/avatar/e64c7d89f26bd1972efa854d13d7dd61?s=80&d=identicon",
        "authorName": "Administrator",
        "id": "gid://gitlab/CommitPresenter/0cb48dd5deddb7632fd7c3defb16075fc6c3ca74",
        "message": "Update .gitlab-ci.yml file",
        "shortId": "0cb48dd5",
        "webUrl": "http://gdk.test:3000/gitlab-org/pipelinestest/-/commit/0cb48dd5deddb7632fd7c3defb16075fc6c3ca74",
      },
      "isLast": false,
    },
    "name": undefined,
    "retryUrl": "null/retry",
  },
  "status": "success",
  "triggerer": {
    "avatarUrl": "/uploads/-/system/user/avatar/1/avatar.png",
    "id": "gid://gitlab/User/1",
    "name": "Administrator",
    "webUrl": "http://gdk.test:3000/root",
  },
  "webPath": "",
}
`;

exports[`deployment_data_transformation_helper convertToDeploymentTableRow should be converted to proper table row data 4`] = `
{
  "actions": [],
  "commit": {
    "author": {
      "avatar_url": "/uploads/-/system/user/avatar/1/avatar.png",
      "path": "http://gdk.test:3000/root",
      "username": "Administrator",
    },
    "commitRef": {
      "name": "main",
    },
    "commitUrl": "http://gdk.test:3000/gitlab-org/pipelinestest/-/commit/0cb48dd5deddb7632fd7c3defb16075fc6c3ca74",
    "shortSha": "0cb48dd5",
    "tag": false,
    "title": "Update .gitlab-ci.yml file",
  },
  "created": "2022-10-17T07:44:17Z",
  "deploymentApproval": {
    "isApprovalActionAvailable": false,
  },
  "finished": "2022-10-17T07:44:43Z",
  "id": "31",
  "job": undefined,
  "rollback": null,
  "status": "success",
  "triggerer": {
    "avatarUrl": "/uploads/-/system/user/avatar/1/avatar.png",
    "id": "gid://gitlab/User/1",
    "name": "Administrator",
    "webUrl": "http://gdk.test:3000/root",
  },
  "webPath": "",
}
`;

exports[`deployment_data_transformation_helper convertToDeploymentTableRow should be converted to proper table row data 5`] = `
{
  "actions": [],
  "commit": {
    "author": {
      "avatar_url": "/uploads/-/system/user/avatar/1/avatar.png",
      "path": "http://gdk.test:3000/root",
      "username": "Administrator",
    },
    "commitRef": {
      "name": "main",
    },
    "commitUrl": "http://gdk.test:3000/gitlab-org/pipelinestest/-/commit/0cb48dd5deddb7632fd7c3defb16075fc6c3ca74",
    "shortSha": "0cb48dd5",
    "tag": false,
    "title": "Update .gitlab-ci.yml file",
  },
  "created": "2022-10-17T07:44:17Z",
  "deploymentApproval": {
    "isApprovalActionAvailable": false,
  },
  "finished": "",
  "id": "31",
  "job": null,
  "rollback": null,
  "status": "success",
  "triggerer": {
    "avatarUrl": "/uploads/-/system/user/avatar/1/avatar.png",
    "id": "gid://gitlab/User/1",
    "name": "Administrator",
    "webUrl": "http://gdk.test:3000/root",
  },
  "webPath": "",
}
`;

exports[`deployment_data_transformation_helper getAuthorFromCommit should be properly converted 1`] = `
{
  "avatar_url": "/uploads/-/system/user/avatar/1/avatar.png",
  "path": "http://gdk.test:3000/root",
  "username": "Administrator",
}
`;

exports[`deployment_data_transformation_helper getAuthorFromCommit should be properly converted 2`] = `
{
  "avatar_url": "https://www.gravatar.com/avatar/91811aee1dec1b2655fa56f894e9e7c9?s=80&d=identicon",
  "path": "mailto:azubov@gitlab.com",
  "username": "Andrei Zubov",
}
`;

exports[`deployment_data_transformation_helper getCommitFromDeploymentNode should get correclty formatted commit object 1`] = `
{
  "author": {
    "avatar_url": "/uploads/-/system/user/avatar/1/avatar.png",
    "path": "http://gdk.test:3000/root",
    "username": "Administrator",
  },
  "commitRef": {
    "name": "main",
  },
  "commitUrl": "http://gdk.test:3000/gitlab-org/pipelinestest/-/commit/0cb48dd5deddb7632fd7c3defb16075fc6c3ca74",
  "shortSha": "0cb48dd5",
  "tag": false,
  "title": "Update .gitlab-ci.yml file",
}
`;
